File: TypeScript/app/handler/NormalItemHandler.ts
import { Item } from '../gilded-rose';
import { ItemHandler } from './ItemHandler';
import { ItemUtils } from '../utility/ItemUtils';

export class NormalItemHandler implements ItemHandler {
  handles(item: Item): boolean {
    return false;
  }

  updateItemQuality(item: Item): void {
    if (item.quality > 0) {
      ItemUtils.decreaseQuality(item);
    }

    ItemUtils.decreaseSellIn(item);

    if (ItemUtils.shouldDegradeTwice(item)) {
      ItemUtils.decreaseQuality(item);
    }
  }
}

File: Java/src/main/java/com/gildedrose/handler/NormalItemHandler.java
package com.gildedrose.handler;

import com.gildedrose.Item;
import com.gildedrose.utility.ItemUtilities;

public class NormalItemHandler implements ItemHandler {
    @Override
    public boolean handles(String itemName) {
        return true;
    }

    @Override
    public void updateItem(Item item) {
        if (ItemUtilities.canDecreaseQuality(item)) {
            ItemUtilities.decreaseQuality(item);
        }

        ItemUtilities.decreaseSellIn(item);

        if (ItemUtilities.isExpired(item) && ItemUtilities.canDecreaseQuality(item)) {
            ItemUtilities.decreaseQuality(item);
        }
    }
}
